File: Leetcode/Number of Music Playlists/Number of Music Playlists.py
# Your music player contains N different songs and she wants to listen to L (not necessarily different) songs during your trip.  
# You create a playlist so that:

# Every song is played at least once
# A song can only be played again only if K other songs have been played
# Return the number of possible playlists.  As the answer can be very large, return it modulo 10^9 + 7.



from functools import lru_cache

class Solution:
    def numMusicPlaylists(self, N, L, K):
        @lru_cache(None)
        # dp[i][j] be the number of playlists of length i that have exactly j unique songs.
        def dp(i, j):
            if i == 0:
                return +(j == 0)
                # Last song, we either played a song for the first time or we didn't. If we did, then we had dp[i - 1][j - 1] * (N - j + 1) 
                # ways to choose it. If we didn't, then we repeated a previous song in dp[i-1][j] * max(j-K, 0) ways
            ans = dp(i-1, j-1) * (N-j+1)
            ans += dp(i-1, j) * max(j-K, 0)
            return ans % (10**9+7)

        return dp(L, N)

File: Leetcode/Number of Music Playlists/Number of Music Playlists.java
class Solution {
    //defining mod value
    long mod = (long)1e9 + 7;
    public int numMusicPlaylists(int N, int L, int K) {
        //Creating dp array
        long[][] dp = new long[N + 1][L + 1];
        
        //Looping from K+1 to N different songs
        for (int i = K + 1; i <= N; ++i){
            //Looping from value of i and the L songs she wants to listen to
            for (int j = i; j <= L; ++j){
                //If their values are equal or if i is K+1, then number of playlists will be factorial of i
                if ((i == j) || (i == K + 1)){
                    dp[i][j] = factorial(i);
                }
                //Otherwise we can put any song at the end of music list,but it should be different from K last song
                else{
                    dp[i][j] = (dp[i - 1][j - 1] * i + dp[i][j - 1] * (i - K))%mod;
                }
            }
        }
        //Returning the ans
        return (int) dp[N][L];
    }

    //Factorial function that calculates factorial of a number
    long factorial(int n) {
        return n > 0 ? factorial(n - 1) * n % mod : 1;
    }
}
